### ISSUE TITLE
Empty "globals" property causing an error in v3.0.0

### ISSUE THREAD

**Issue by paul-barry-kenzan** [eslint/eslint#7967]
**Tell us about your environment**

* **ESLint Version:** 3.14.0
* **Node Version:** 5.7.0
* **npm Version:** 3.3.8

**What parser (default, Babel-ESLint, etc.) are you using?** None.

**Please show your full configuration:**

```
{
    "env": {
        "browser": true,
        "node": true
    },

    "globals": {},

    "plugins": [],
    "rules": {
        "comma-dangle": [2,"never"],
        "no-cond-assign": 2,
        "no-console": 2,
        "no-constant-condition": 2,
        "no-control-regex": 0,
        "no-debugger": 0,
        "no-dupe-args": 0,
        "no-dupe-keys": 0,
        "no-duplicate-case": 2,
        "no-empty-character-class": 0,
        "no-empty": 2,
        "no-ex-assign": 2,
        "no-extra-boolean-cast": 2,
        "no-extra-parens": [2, "all", {
            "conditionalAssign": false,
            "returnAssign": false,
            "nestedBinaryExpressions": false
        }],
        "no-extra-semi": 2,
        "no-func-assign": 2,
        "no-inner-declarations": 2,
        "no-invalid-regexp": 0,
        "no-irregular-whitespace": 2,
        "no-negated-in-lhs": 2,
        "no-obj-calls": 0,
        "no-regex-spaces": 0,
        "no-sparse-arrays": 2,
        "no-unreachable": 2,
        "use-isnan": 2,
        "valid-jsdoc": 0,
        "valid-typeof": 0,
        "no-unexpected-multiline": 0,
        "accessor-pairs": 2,
        "block-scoped-var": 2,
        "complexity": 2,
        "consistent-return": 0,
        "curly": 2,
        "default-case": 2,
        "dot-notation": 2,
        "dot-location": 0,
        "eqeqeq": [2,"allow-null"],
        "guard-for-in": 0,
        "no-alert": 0,
        "no-caller": 0,
        "no-div-regex": 0,
        "no-else-return": 0,
        "no-empty-label": 0,
        "no-eq-null": 0,
        "no-eval": 2,
        "no-extend-native": 0,
        "no-extra-bind": 2,
        "no-fallthrough": 2,
        "no-floating-decimal": 2,
        "no-implicit-coercion": [2,{"number":true,"string":true,"boolean":false}],
        "no-implied-eval": 2,
        "no-invalid-this": 0,
        "no-iterator": 2,
        "no-labels": 0,
        "no-lone-blocks": 0,
        "no-loop-func": 2,
        "no-multi-spaces": 2,
        "no-multi-str": 0,
        "no-native-reassign": 0,
        "no-new-func": 0,
        "no-new-wrappers": 2,
        "no-new": 2,
        "no-octal-escape": 0,
        "no-octal": 0,
        "no-param-reassign": 0,
        "no-process-env": 0,
        "no-proto": 0,
        "no-redeclare": 2,
        "no-return-assign": 0,
        "no-script-url": 0,
        "no-self-compare": 0,
        "no-sequences": 0,
        "no-throw-literal": 2,
        "no-unused-expressions": 0,
        "no-useless-call": 0,
        "no-void": 0,
        "no-warning-comments": [2,{"terms":["todo"," fixme"," TODO"," FIXME"],"location":"anywhere"}],
        "no-with": 0,
        "radix": 2,
        "vars-on-top": 2,
        "wrap-iife": [2,"inside"],
        "yoda": 0,
        "strict": [2,"global"],
        "init-declarations": 0,
        "no-catch-shadow": 2,
        "no-delete-var": 2,
        "no-label-var": 0,
        "no-shadow-restricted-names": 0,
        "no-shadow": 0,
        "no-undef-init": 0,
        "no-undef": 0,
        "no-undefined": 0,
        "no-unused-vars": 2,
        "no-use-before-define": 0,
        "callback-return": 0,
        "handle-callback-err": 2,
        "no-mixed-requires": 0,
        "no-new-require": 0,
        "no-path-concat": 2,
        "no-process-exit": 2,
        "no-restricted-modules": 0,
        "no-sync": 0,
        "array-bracket-spacing": 2,
        "brace-style": [2,"1tbs",{"allowSingleLine":true}],
        "camelcase": 2,
        "comma-spacing": 2,
        "comma-style": [2,"last"],
        "computed-property-spacing": 0,
        "consistent-this": [0, "self", "that"],
        "eol-last": 0,
        "func-names": 0,
        "func-style": 0,
        "id-length": 0,
        "indent": [2, 2, {
            "VariableDeclarator": 1,
            "SwitchCase": 1
        }],
        "key-spacing": [2,{"beforeColon":false,"afterColon":true}],
        "lines-around-comment": 0,
        "linebreak-style": 0,
        "max-nested-callbacks": [2, {"maximum":8}],
        "new-cap": 2,
        "new-parens": 2,
        "newline-after-var": 2,
        "no-array-constructor": 2,
        "no-continue": 0,
        "no-inline-comments": 0,
        "no-lonely-if": 0,
        "no-mixed-spaces-and-tabs": [2,"smart-tabs"],
        "no-multiple-empty-lines": [2,{"max":1}],
        "no-nested-ternary": 0,
        "no-new-object": 2,
        "no-spaced-func": 2,
        "no-ternary": 0,
        "no-trailing-spaces": 0,
        "no-underscore-dangle": [2, { "allowAfterThis": true }],
        "no-unneeded-ternary": 2,
        "object-curly-spacing": [2,"always",{}],
        "one-var": 0,
        "operator-assignment": 0,
        "operator-linebreak": 0,
        "padded-blocks": [2, {"switches":"always"}],
        "quote-props": [2, "consistent"],
        "quotes": [2,"single","avoid-escape"],
        "id-match": 0,
        "semi-spacing": [2,{"after":true}],
        "semi": [2,"always"],
        "sort-vars": 0,
        "keyword-spacing": 2,
        "space-before-blocks": 2,
        "space-before-function-paren": 0,
        "space-in-parens": 2,
        "space-infix-ops": 2,
        "space-return-throw-case": 0,
        "space-unary-ops": 0,
        "spaced-comment": [2, "always", {
            "line": {
                "markers": ["/"],
                "exceptions": ["-", "+"]
            },
            "block": {
                "markers": ["!"],
                "exceptions": ["*"]
            }
        }],
        "wrap-regex": 2,
        "arrow-parens": 0,
        "arrow-spacing": 0,
        "constructor-super": 0,
        "generator-star-spacing": 0,
        "no-class-assign": 0,
        "no-const-assign": 0,
        "no-this-before-super": 0,
        "no-var": 0,
        "object-shorthand": 0,
        "prefer-const": 0,
        "prefer-spread": 0,
        "prefer-reflect": 0,
        "require-yield": 0,
        "max-depth": [2,4],
        "max-len": [2,200,1,{"ignorePattern":"true"}],
        "max-params": 0,
        "max-statements": 0,
        "no-bitwise": [2, { "allow": ["~"] }],
        "no-plusplus": 2
    }
}
```

**What did you do? Please include the actual source code causing the issue.**
We updated `gulp-eslint` from `2.0.0` to `3.0.1`, which internally updated `ESLint` from `^2.0.0` to `^3.0.0`. Our configuration remained the same, however, we're now receiving the error described below.

**What did you expect to happen?**
With the exception of the breaking changes mentioned in the Migration to 3.0.0 guide, we expected to be able to update our dependencies with ease. 

**What actually happened? Please include the actual, raw output from ESLint.**
When attempting to run the same, unchanged task as we have for weeks/months, we now receive the following error which pinpoints the `globals` property being an issue. By chance, I decided to remove the empty `globals` property and was able to successfully run the lint task. Is there any information why an empty `globals` property throws an error? Is it possible to have such information added to ESLint documentation somewhere?

Additional discussions around GitHub:
 - https://github.com/adametry/gulp-eslint/issues/177
 - https://github.com/eslint/eslint/pull/6922

Neither of the aforementioned discussions mentions a solution. 

```
/Users/path/to/project/directory/node_modules/eslint/lib/config.js:214
        this.globals = (options.globals || []).reduce((globals, def) => {
                                               ^

TypeError: (options.globals || []).reduce is not a function
    at new Config (/Users/paulbarry/Projects/kenzan/kenzanlabs/pipeline-validate-html/node_modules/eslint/lib/config.js:214:48)
    at CLIEngine.executeOnText (/Users/paulbarry/Projects/kenzan/kenzanlabs/pipeline-validate-html/node_modules/eslint/lib/cli-engine.js:741:28)
    at verify (/Users/paulbarry/Projects/kenzan/kenzanlabs/pipeline-validate-html/node_modules/gulp-eslint/index.js:20:25)
    at Transform._transform (/Users/paulbarry/Projects/kenzan/kenzanlabs/pipeline-validate-html/node_modules/gulp-eslint/index.js:70:17)
    at Transform._read (_stream_transform.js:167:10)
    at Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:301:12)
    at writeOrBuffer (_stream_writable.js:287:5)
    at Transform.Writable.write (_stream_writable.js:215:11)
    at Stream.method [as write] (/Users/paulbarry/Projects/kenzan/kenzanlabs/pipeline-validate-html/node_modules/duplexer/index.js:47:39)

```

**Comment by platinumazure** [comment]
Might be a bug. It's a bit interesting that we expect an object in configuration files but are expecting an array in the line of code that throws-- perhaps our conversion logic doesn't handle empty objects.

**Comment by ilyavolodin** [comment]
For some reason I can't reproduce this in the online demo. @paul-barry-kenzan Could you try to create a short repro case here: http://eslint.org/demo/
Based on the code that's erroring out I'm not sure how this is happening. Since it checks if globals are provided and if not, uses an empty array.

**Comment by platinumazure** [comment]
@ilyavolodin You can specify globals via inline config in the demo, but not via a config file. You might need to try with a config file (seems an empty object of globals might be enough to repro).

Unfortunately I'm stuck at work, otherwise I would be trying to repro this.

**Comment by paul-barry-kenzan** [comment]
@ilyavolodin @platinumazure 
Thanks for the quick response. 

I'm not able to mimic my setup simply because the demo app doesn't provide a means of setting "globals". 

After looking at the code a bit more, from what it looks like the check and subsequent creation of the empty array is incorrect. If I provide `options.globals = {}`, the condition _will not_ create the empty array, since `{} !== "undefined"`, thus, it attempts to use `.reduce` on the empty object.

I suggest updating the condition to check  something along the lines of:
```javascript
this.globals = (Object.keys(options.globals).length > 0 ? options.globals : []).reduce((globals, def) => {
```

Please let me know your thoughts.

**Comment by nzakas** [comment]
Can you try this setup using ESLint directly (outside of Gulp) to see if you get the same result? If not, it could be a problem with gulp-eslint passing configuration in.

**Comment by paul-barry-kenzan** [comment]
@platinumazure, @nzakas, @ilyavolodin 

I've done some digging and I think I found the cause, and a potentially solution. 

The CLI accepts `--global foo, bar:true`, which is parsed as `[ 'foo', 'bar:true ]` -- an array of strings, which are processed and turned into an object.

ESLint also accepts a config object which includes `{ globals: { foo: false, bar: true } }` -- a collection of key:value pairs. This is where the problem occurs. The Config class expects the `options.global` value to be an array of strings. However, since the `.eslintrc` file declares this value as an object, the attempt to use `.reduce()` fails (hence the `options.globals || []` line mentioned in the error above).

The potentially solution would be to pass through the `options.globals` object if it were already in an acceptable format, and leave the parsing of the array of strings for only when the CLI is utilized. 

Do you feel this is a viable solution? I will gladly submit a PR with my attempt if you feel the approach is sufficient. 

Thanks.

**Comment by nzakas** [comment]
It's still unclear to me if this is a gulp-eslint issue or not. Can you please confirm if you see the same problem using ESLint directly?

**Comment by paul-barry-kenzan** [comment]
@nzakas 
I'm having a hard time replicating the error from the command-line. This forced me to trace the path through our usage.

Here's a bit more info on our use case.

We've created a pipeline that wraps `gulp-eslint`, which uses an opinionated `.eslintrc` for our internal projects when linting files, eliminating the need to create a config per project. This config file is parsed to JSON and is passed through to `gulp-eslint` as an object. `gulp-eslint` accepts the options properly, and passes them along to `ESLint` via a [call to `new CLIEngine(options)`](https://github.com/adametry/gulp-eslint/blob/master/index.js#L17).

This is where things get strange. 

The CLIEngine [immediately joins the options passed in with the default CLI options](https://github.com/eslint/eslint/blob/master/lib/cli-engine.js#L442), which the `globals` property seems to be inconsistent with the documentation.

```javascript
// from conf/cli-options.js
module.exports = {
  ...
  globals: [], // <- notice Array
  ...
}
```
While [the documentation for `.eslintrc` files suggests an object](http://eslint.org/docs/user-guide/configuring#specifying-globals)

```javascript
{
    "globals": {
        "var1": true,
        "var2": false
    }
}
```

One other piece of the puzzle, that certainly falls on `gulp-eslint`, is that the documentation there suggests that [the `options.globals` property be an array](https://github.com/adametry/gulp-eslint#optionsglobals).

I await your suggestions on how to proceed. 

1. Does the ESLint documentation need to be updated to indicate an Array for globals?
1. Does the `gulp-eslint` documentation need to be updated to indicate an Object should be passed? (Obviously not the responsibility of this project team)
1. If an object should be passed to ESLint, does my PR provide proper support for empty objects?

**Comment by mysticatea** [comment]
I think this is not a bug. CLIEngine's `globals` option is an array, as documented.
http://eslint.org/docs/developer-guide/nodejs-api#cliengine

**Comment by paul-barry-kenzan** [comment]
@mysticatea 
Agreed that the CLI option is an array. However, I'm not using the CLI directly and using a `.eslintrc` file, which is [documented as requiring an object](http://eslint.org/docs/user-guide/configuring#specifying-globals). That's really part of the confusion. The underlying CLI is requiring an array, but the docs suggest providing an object.

**Comment by mysticatea** [comment]
I could not reproduce that.

![image](https://cloud.githubusercontent.com/assets/1937871/22528767/0b31935a-e918-11e6-8717-dfc6b4fe056b.png)

Could you give me repro step without gulp?

**Comment by cecilemuller** [comment]
Why is CLIEngine expecting an Array for `globals` when .eslintrc expect an Object ? It would be less confusing if both used the same type.

**Comment by mysticatea** [comment]
I'm sorry, I'm not sure since it had been the current form already when I joined...

**Comment by kaicataldo** [comment]
Maybe @nzakas can shed some light on what seems like an inconsistency.

Seems to me like we should either make this clearer in the documentation or update it so that the CLI accepts the same data type as configuration files. I think this could be done in a backwards compatible manner so we don't break things. Perhaps deprecating in the next major and removing it in the major after that.

**Comment by nzakas** [comment]
First, it's important to understand that the object CLIEngine accepts as a parameter is explicitly *not* a config object even though there are similar keys. This is intentional and not a bug, though this has been a point of confusion. Most users are just using configs, it's only tool writers that need to worry about CLIEngine options. 

IIRC the initial implementation of the command line flag was just a way to add global definitions that were writable, and that was partly because we were using a simple CLI argument parser that didn't understand anything other than strings. So our choices were to add something to scratch a user itch immediately or else spend a bunch of time trying to figure out the right way to represent key-value pairs on the command line. We did the former.

While I'm not opposed to making a change as long as it's backwards compatible, this still seems like it's a problem with gulp-eslint calling CLIEngine directly. We haven't seen any evidence of the problem just using the ESLint CLI, and so making a change there without understanding the true root cause of the integration failure doesn't seem like the best path forward to me. We would just be fixing around the problem.

**Comment by gyandeeps** [comment]
@eslint/eslint-team should this be added to TSC agenda so that we can decided if we are going ahead with a change or not?

**Comment by not-an-aardvark** [comment]
I don't think we ever figured out what the problem was because we couldn't replicate it from the command line, so we can't go ahead with any change yet.

That said, I suspect this was caused by https://github.com/eslint/eslint/issues/8329, so it should be fixed in ESLint v3.19.0. @paul-barry-kenzan Does this issue still occur when you use the latest version of ESLint?

**Comment by not-an-aardvark** [comment]
Closing because there was no response from the author and I suspect this has been fixed by https://github.com/eslint/eslint/issues/8329. Feel free to let me know if this is not the case so that I can reopen.

**Comment by matthew-dean** [comment]
@not-an-aardvark This is definitely still an issue in the latest gulp-eslint. But the issue is that gulp-eslint cannot just pass the `.eslintrc` object to the ESLint JS API. The `.eslintrc` still specifies that globals (and other options) should be defined as objects, but the API expects arrays. This issue is incorrectly closed. I don't see any evidence of a fix. See: https://github.com/adametry/gulp-eslint/issues/177

Please re-open.

**Comment by matthew-dean** [comment]
If I change my `.eslintrc` values to an array, I now get two errors:
```
TypeError: (options.globals || []).reduce is not a function
    at new Config (.../eslint/lib/config.js:94:48)

Error: ESLint configuration in /Users/madean/git/frontend-eacom/.eslintrc is invalid:
        - Property "env" is the wrong type (expected object but got `["node","browser"]`).

    at validateConfigSchema (.../eslint/lib/config/config-validator.js:214:15)
```
So, the CLIEngine expects an array, but an array can't actually be passed via the Node.js API (if just reading the `.eslintrc` file as an object).

**Comment by not-an-aardvark** [comment]
>  But the issue is that gulp-eslint cannot just pass the `.eslintrc` object to the ESLint JS API. 

I'm confused. Could you clarify how gulp-eslint is using the API? I'm not sure passing an `.eslintrc` object directly to `CLIEngine` is supposed to be supported.

**Comment by shinnn** [comment]
> I'm not sure passing an .eslintrc object directly to CLIEngine is supposed to be supported.

It's not supported. gulp-eslint accepts the same options as `CLIEngine` does.

**Comment by platinumazure** [comment]
Should this be closed? gulp-eslint supports the CLIEngine options and passes them to CLIEngine-- there is no support in either CLIEngine or gulp-eslint for trying to pass a config object directly.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
